managed/services/management/agent.go

```diff
@@ -0,0 +1,175 @@
+
+package management
+
+import (
+	"context"
+
+	"github.com/AlekSi/pointer"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm/api/managementpb"
+	"github.com/percona/pmm/managed/models"
+)
+
+// AgentService represents service for working with agents.
+type AgentService struct {
+	db    *reform.DB
+	r     agentsRegistry
+	state agentsStateUpdater
+	vmdb  prometheusService
+
+	managementpb.UnimplementedAgentServer
+}
+
+// NewAgentService creates AgentService instance.
+func NewAgentService(db *reform.DB, r agentsRegistry, state agentsStateUpdater, vmdb prometheusService) *AgentService {
+	return &AgentService{
+		db:    db,
+		r:     r,
+		state: state,
+		vmdb:  vmdb,
+	}
+}
+
+// ListAgents returns a filtered list of Agents.
+//
+//nolint:unparam
+func (s *AgentService) ListAgents(ctx context.Context, req *managementpb.ListAgentRequest) (*managementpb.ListAgentResponse, error) {
+	if req.ServiceId == "" {
+		return nil, status.Error(codes.InvalidArgument, "service_id is required")
+	}
+
+	serviceID := req.GetServiceId()
+
+	var agents []*models.Agent
+	var service *models.Service
+
+	e := s.db.InTransaction(func(tx *reform.TX) error {
+		var err error
+
+		agents, err = models.FindAgents(tx.Querier, models.AgentFilters{})
+		if err != nil {
+			return err
+		}
+
+		service, err = models.FindServiceByID(tx.Querier, serviceID)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	if e != nil {
+		return nil, e
+	}
+
+	svcAgents := []*managementpb.GenericAgent{}
+
+	for _, agent := range agents {
+		// agent is not an exporter, but it runs on the same node as the service (p.e. pmm-agent)
+		if agent.ServiceID == nil && pointer.GetString(agent.RunsOnNodeID) == service.NodeID {
+			ag, err := s.toAPIAgent(agent)
+			if err != nil {
+				return nil, err
+			}
+			svcAgents = append(svcAgents, ag)
+		}
+
+		// vmagent that runs on the same node as the service
+		if pointer.GetString(agent.NodeID) == service.NodeID && agent.AgentType == models.VMAgentType {
+			ag, err := s.toAPIAgent(agent)
+			if err != nil {
+				return nil, err
+			}
+			svcAgents = append(svcAgents, ag)
+		}
+
+		// agent is an exporter for this service
+		if agent.ServiceID != nil && pointer.GetString(agent.ServiceID) == serviceID {
+			ag, err := s.toAPIAgent(agent)
+			if err != nil {
+				return nil, err
+			}
+			svcAgents = append(svcAgents, ag)
+		}
+	}
+
+	res := &managementpb.ListAgentResponse{
+		Agents: svcAgents,
+	}
+
+	return res, nil
+}
+
+func (s *AgentService) toAPIAgent(agent *models.Agent) (*managementpb.GenericAgent, error) {
+	const pass = "**********"
+
+	labels, err := agent.GetCustomLabels()
+	if err != nil {
+		return nil, err
+	}
+
+	ga := &managementpb.GenericAgent{
+		AgentId:                        agent.AgentID,
+		AgentType:                      string(agent.AgentType),
+		AwsAccessKey:                   pointer.GetString(agent.AWSAccessKey),
```

Comment:
If we treat it similar to the `username`, then it's useful to leave it as is. The problem is that the user is sometimes wondering which key is being used for this agent, since there can be many access keys issued for the same object AWS. However, the AwsSecretKey is treated as a secret.

Reply:
I ended up using fields a-la `bool is_password_set` to prevent us from unintentionally leaking secrets.